### ISSUE TITLE
Large amount of S3 calls of compactor in small Thanos instance

### ISSUE THREAD

**Issue by albgus** [thanos-io/thanos#6033]
**Thanos, Prometheus and Golang version used**:
Installed by Helm chart: `bitnami/thanos-11.6.8`
<details>

```
#values.yml
existingObjstoreSecret: thanos-objstore
clusterDomain: cluster.local

query:
  enabled: true
  replicaLabel:
    - replica
    - receive

queryFrontend:
  enabled: true

compactor:
  enabled: true
  retentionResolutionRaw: 0s
  retentionResolution5m: 0s
  retentionResolution1h: 0s
  persistence:
    enabled: false

storegateway:
  enabled: true
  persistence:
    enabled: false

ruler:
  enabled: false

receive:
  enabled: true
  mode: standalone
  tsdbRetention: 30d
  replicationFactor: 1
  persistence:
    enabled: true
    existingClaim: thanos-receive

receiveDistributor:
  enabled: false

metrics:
  enabled: true

minio:
  enabled: false
```

```
thanos, version 0.30.1 (branch: HEAD, revision: 27e4714174e2b7d4542ce9a723716134afba1579)
  build user:       root@90643487b1ff
  build date:       20230104-18:37:40
  go version:       go1.19.4
  platform:         linux/amd64
```

```
Containers:
  compactor:
    Container ID:  containerd://1b2c5ad8552d156f6947ba691490404cd6718c1b16bac6dbac706fef768185b2
    Image:         docker.io/bitnami/thanos:0.30.1-scratch-r0
    Image ID:      docker.io/bitnami/thanos@sha256:c35abdf94a0d111a2e710b08e47a41d4a5ee58ac36a545192c44fdf2edb212ec
    Port:          10902/TCP
    Host Port:     0/TCP
    Args:      compact
      --log.level=info
      --log.format=logfmt
      --http-address=0.0.0.0:10902
      --data-dir=/data
      --retention.resolution-raw=0s
      --retention.resolution-5m=0s
      --retention.resolution-1h=0s
      --consistency-delay=30m
      --objstore.config-file=/conf/objstore.yml
      --wait
```
</details>

**Object Storage Provider**: AWS S3

**What happened**:
I am seeing a very large amount of requests billed as `Amazon Simple Storage Service EUN1-Requests-Tier2` in a test account with a small Thanos installation. I'm seeing about 40 Million requests per month, which comes down to around 60,000 requests per hour, or 16-20 requests / second.

**What you expected to happen**:
I would have expected a small instance to perform significantly less requests against S3.

This is with a very small instance of Thanos, the S3 bucket having 1.5GB of storage in about 550 files, containing a bit over 2 years of data at fairly low resolution (1-5 minutes)

**Full logs to relevant components**:

Looks like almost all of the requests comes from the compactor, in an about 50/50 mix between `HEAD [...]/index.json`, or GET requests for `deletion-mark.json` or `no-compact-mark.json`
```
$ wc -l s3.log
1 338 023 s3.log
$ grep -c 'HEAD.*/meta.json' s3.log
649 613
$ grep -c 'NoSuchKey' s3.log
610 258
```

Example logs from S3
```
thanos.i.example.com [09/Jan/2023:21:46:22 +0000] 1.2.3.4 arn:aws:iam::012345678910:user/test-infra-thanos XCJRGVE9EF1QQVQC REST.GET.OBJECT test-infra/01GJG6H63WAJ3CJ8WH20SYVK5R/deletion-mark.json "GET /thanos.i.example.com/test-infra/01GJG6H63WAJ3CJ8WH20SYVK5R/deletion-mark.json HTTP/1.1" 404 NoSuchKey 327 - 7 - "-" "MinIO (linux; amd64) minio-go/v7.0.45 thanos-compact/0.30.1 (go1.19.4)"
thanos.i.example.com [09/Jan/2023:21:46:22 +0000] 1.2.3.4 arn:aws:iam::012345678910:user/test-infra-thanos XCJT5KFN7GSKH45Z REST.GET.OBJECT test-infra/01F6G9S4JBT0YXB6N22WW1845C/deletion-mark.json "GET /thanos.i.example.com/test-infra/01F6G9S4JBT0YXB6N22WW1845C/deletion-mark.json HTTP/1.1" 404 NoSuchKey 327 - 7 - "-" "MinIO (linux; amd64) minio-go/v7.0.45 thanos-compact/0.30.1 (go1.19.4)"
thanos.i.example.com [09/Jan/2023:21:46:22 +0000] 1.2.3.4 arn:aws:iam::012345678910:user/test-infra-thanos XCJTQMSGR2R7QDTH REST.GET.OBJECT test-infra/01GJJB5XCCB5ZMGHSSVZDFWVD3/deletion-mark.json "GET /thanos.i.example.com/test-infra/01GJJB5XCCB5ZMGHSSVZDFWVD3/deletion-mark.json HTTP/1.1" 404 NoSuchKey 327 - 7 - "-" "MinIO (linux; amd64) minio-go/v7.0.45 thanos-compact/0.30.1 (go1.19.4)"
thanos.i.example.com [09/Jan/2023:21:56:24 +0000] 1.2.3.4 arn:aws:iam::012345678910:user/test-infra-thanos 2QR25GBJDTZGA6J4 REST.GET.OBJECT test-infra/01GJG6NQHP1GCXYV9TGMNESC3D/no-compact-mark.json "GET /thanos.i.example.com/test-infra/01GJG6NQHP1GCXYV9TGMNESC3D/no-compact-mark.json HTTP/1.1" 404 NoSuchKey 329 - 12 - "-" "MinIO (linux; amd64) minio-go/v7.0.45 thanos-compact/0.30.1 (go1.19.4)"
thanos.i.example.com [09/Jan/2023:21:56:24 +0000] 1.2.3.4 arn:aws:iam::012345678910:user/test-infra-thanos 2QR2ASW479QYA3R7 REST.GET.OBJECT test-infra/01GPC73YN26C7893Q3VAA28D6B/no-compact-mark.json "GET /thanos.i.example.com/test-infra/01GPC73YN26C7893Q3VAA28D6B/no-compact-mark.json HTTP/1.1" 404 NoSuchKey 329 - 12 - "-" "MinIO (linux; amd64) minio-go/v7.0.45 thanos-compact/0.30.1 (go1.19.4)"
thanos.i.example.com [09/Jan/2023:21:56:24 +0000] 1.2.3.4 arn:aws:iam::012345678910:user/test-infra-thanos 2QR2C2NVJDEBKRMZ REST.GET.OBJECT test-infra/01GJG4K4480ES027J2KC4SFV7J/deletion-mark.json "GET /thanos.i.example.com/test-infra/01GJG4K4480ES027J2KC4SFV7J/deletion-mark.json HTTP/1.1" 404 NoSuchKey 327 - 8 - "-" "MinIO (linux; amd64) minio-go/v7.0.45 thanos-compact/0.30.1 (go1.19.4)"
thanos.i.example.com [09/Jan/2023:20:59:54 +0000] 1.2.3.4 arn:aws:iam::012345678910:user/test-infra-thanos 65Z1XC39562G1968 REST.HEAD.OBJECT test-infra/01F9RBZ8KX4DFZA43M8BHY62JC/meta.json "HEAD /thanos.i.example.com/test-infra/01F9RBZ8KX4DFZA43M8BHY62JC/meta.json HTTP/1.1" 200 - - 9907 9 - "-" "MinIO (linux; amd64) minio-go/v7.0.45 thanos-compact/0.30.1 (go1.19.4)"
thanos.i.example.com [09/Jan/2023:20:59:54 +0000] 1.2.3.4 arn:aws:iam::012345678910:user/test-infra-thanos 65Z22HFBACCH2ADC REST.HEAD.OBJECT test-infra/01GGW00S98GF4DZZE580RPDZXF/meta.json "HEAD /thanos.i.example.com/test-infra/01GGW00S98GF4DZZE580RPDZXF/meta.json HTTP/1.1" 200 - - 9045 17 - "-" "MinIO (linux; amd64) minio-go/v7.0.45 thanos-compact/0.30.1 (go1.19.4)"
thanos.i.example.com [09/Jan/2023:20:59:54 +0000] 1.2.3.4 arn:aws:iam::012345678910:user/test-infra-thanos 65Z2PV1CY6R20BNN REST.HEAD.OBJECT test-infra/01GP9C2JW7BFPVGRB20JCBW8HE/meta.json "HEAD /thanos.i.example.com/test-infra/01GP9C2JW7BFPVGRB20JCBW8HE/meta.json HTTP/1.1" 200 - - 691 13 - "-" "MinIO (linux; amd64) minio-go/v7.0.45 thanos-compact/0.30.1 (go1.19.4)"
thanos.i.example.com [09/Jan/2023:20:59:54 +0000] 1.2.3.4 arn:aws:iam::012345678910:user/test-infra-thanos 65Z2RFKHSXGP9KCJ REST.HEAD.OBJECT test-infra/01GJG6HYWK0RAQJW524JAJ6YR4/meta.json "HEAD /thanos.i.example.com/test-infra/01GJG6HYWK0RAQJW524JAJ6YR4/meta.json HTTP/1.1" 200 - - 9049 11 - "-" "MinIO (linux; amd64) minio-go/v7.0.45 thanos-compact/0.30.1 (go1.19.4)"
```
<details>Compactor logs
<p>

```
level=info ts=2023-01-09T21:36:22.629309533Z caller=compact.go:1291 msg="start sync of metas"
level=info ts=2023-01-09T21:36:23.085786189Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=456.549601ms duration_ms=456 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:36:23.42711341Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=341.112912ms duration_ms=341 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:36:23.427256864Z caller=clean.go:34 msg="started cleaning of aborted partial uploads"
level=info ts=2023-01-09T21:36:23.427281465Z caller=clean.go:61 msg="cleaning of aborted partial uploads done"
level=info ts=2023-01-09T21:36:23.42729718Z caller=blocks_cleaner.go:44 msg="started cleaning of blocks marked for deletion"
level=info ts=2023-01-09T21:36:23.42733271Z caller=blocks_cleaner.go:58 msg="cleaning of blocks marked for deletion done"
level=info ts=2023-01-09T21:36:23.795860475Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=368.522075ms duration_ms=368 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:36:23.795976484Z caller=compact.go:1296 msg="start of GC"
level=info ts=2023-01-09T21:36:23.796886112Z caller=compact.go:1319 msg="start of compactions"
level=info ts=2023-01-09T21:36:23.797733551Z caller=compact.go:1355 msg="compaction iterations done"
level=info ts=2023-01-09T21:36:23.797798598Z caller=compact.go:430 msg="start first pass of downsampling"
level=info ts=2023-01-09T21:36:24.143234162Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=345.401832ms duration_ms=345 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:36:24.152232175Z caller=compact.go:444 msg="start second pass of downsampling"
level=info ts=2023-01-09T21:36:24.532001531Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=379.733126ms duration_ms=379 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:36:24.54048509Z caller=compact.go:451 msg="downsampling iterations done"
level=info ts=2023-01-09T21:36:24.877461273Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=336.938885ms duration_ms=336 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:36:24.87758081Z caller=retention.go:32 msg="start optional retention"
level=info ts=2023-01-09T21:36:24.877619402Z caller=retention.go:47 msg="optional retention apply done"
level=info ts=2023-01-09T21:36:25.237112881Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=359.471303ms duration_ms=359 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:36:25.237238629Z caller=clean.go:34 msg="started cleaning of aborted partial uploads"
level=info ts=2023-01-09T21:36:25.237260988Z caller=clean.go:61 msg="cleaning of aborted partial uploads done"
level=info ts=2023-01-09T21:36:25.237276747Z caller=blocks_cleaner.go:44 msg="started cleaning of blocks marked for deletion"
level=info ts=2023-01-09T21:36:25.237311343Z caller=blocks_cleaner.go:58 msg="cleaning of blocks marked for deletion done"
level=info ts=2023-01-09T21:36:54.197922581Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=253.098354ms duration_ms=253 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:37:54.206961741Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=262.236529ms duration_ms=262 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:38:54.237132756Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=292.379752ms duration_ms=292 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:39:54.189723553Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=244.982464ms duration_ms=244 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:40:54.281415238Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=336.728403ms duration_ms=336 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:41:22.629232573Z caller=compact.go:1291 msg="start sync of metas"
level=info ts=2023-01-09T21:41:23.113351584Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=484.134113ms duration_ms=484 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:41:23.464824382Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=351.250383ms duration_ms=351 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:41:23.464940114Z caller=clean.go:34 msg="started cleaning of aborted partial uploads"
level=info ts=2023-01-09T21:41:23.464973383Z caller=clean.go:61 msg="cleaning of aborted partial uploads done"
level=info ts=2023-01-09T21:41:23.464988945Z caller=blocks_cleaner.go:44 msg="started cleaning of blocks marked for deletion"
level=info ts=2023-01-09T21:41:23.465023218Z caller=blocks_cleaner.go:58 msg="cleaning of blocks marked for deletion done"
level=info ts=2023-01-09T21:41:23.821646543Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=356.623108ms duration_ms=356 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:41:23.821837441Z caller=compact.go:1296 msg="start of GC"
level=info ts=2023-01-09T21:41:23.822956468Z caller=compact.go:1319 msg="start of compactions"
level=info ts=2023-01-09T21:41:23.824211984Z caller=compact.go:1355 msg="compaction iterations done"
level=info ts=2023-01-09T21:41:23.824288458Z caller=compact.go:430 msg="start first pass of downsampling"
level=info ts=2023-01-09T21:41:24.155779485Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=331.465285ms duration_ms=331 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:41:24.164656225Z caller=compact.go:444 msg="start second pass of downsampling"
level=info ts=2023-01-09T21:41:24.515630779Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=350.937276ms duration_ms=350 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:41:24.523994286Z caller=compact.go:451 msg="downsampling iterations done"
level=info ts=2023-01-09T21:41:24.882721047Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=358.677568ms duration_ms=358 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:41:24.882834518Z caller=retention.go:32 msg="start optional retention"
level=info ts=2023-01-09T21:41:24.882872977Z caller=retention.go:47 msg="optional retention apply done"
level=info ts=2023-01-09T21:41:25.280126209Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=397.23131ms duration_ms=397 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:41:25.280246097Z caller=clean.go:34 msg="started cleaning of aborted partial uploads"
level=info ts=2023-01-09T21:41:25.280269503Z caller=clean.go:61 msg="cleaning of aborted partial uploads done"
level=info ts=2023-01-09T21:41:25.280285496Z caller=blocks_cleaner.go:44 msg="started cleaning of blocks marked for deletion"
level=info ts=2023-01-09T21:41:25.280319508Z caller=blocks_cleaner.go:58 msg="cleaning of blocks marked for deletion done"
level=info ts=2023-01-09T21:41:54.234273488Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=289.38488ms duration_ms=289 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:42:54.172009137Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=227.668901ms duration_ms=227 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:43:54.257231606Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=313.268026ms duration_ms=313 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:44:54.232058255Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=287.306826ms duration_ms=287 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:45:54.187292048Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=242.468381ms duration_ms=242 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:46:22.629135101Z caller=compact.go:1291 msg="start sync of metas"
level=info ts=2023-01-09T21:46:23.085725635Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=456.655381ms duration_ms=456 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:46:23.430794619Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=344.854733ms duration_ms=344 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:46:23.430910434Z caller=clean.go:34 msg="started cleaning of aborted partial uploads"
level=info ts=2023-01-09T21:46:23.430933074Z caller=clean.go:61 msg="cleaning of aborted partial uploads done"
level=info ts=2023-01-09T21:46:23.430948122Z caller=blocks_cleaner.go:44 msg="started cleaning of blocks marked for deletion"
level=info ts=2023-01-09T21:46:23.430986987Z caller=blocks_cleaner.go:58 msg="cleaning of blocks marked for deletion done"
level=info ts=2023-01-09T21:46:23.791041727Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=360.049239ms duration_ms=360 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:46:23.791156348Z caller=compact.go:1296 msg="start of GC"
level=info ts=2023-01-09T21:46:23.792092458Z caller=compact.go:1319 msg="start of compactions"
level=info ts=2023-01-09T21:46:23.79293398Z caller=compact.go:1355 msg="compaction iterations done"
level=info ts=2023-01-09T21:46:23.792995905Z caller=compact.go:430 msg="start first pass of downsampling"
level=info ts=2023-01-09T21:46:24.153357808Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=360.337512ms duration_ms=360 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:46:24.162179161Z caller=compact.go:444 msg="start second pass of downsampling"
level=info ts=2023-01-09T21:46:24.50385105Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=341.635578ms duration_ms=341 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:46:24.512246975Z caller=compact.go:451 msg="downsampling iterations done"
level=info ts=2023-01-09T21:46:24.848390243Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=336.102565ms duration_ms=336 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:46:24.848528041Z caller=retention.go:32 msg="start optional retention"
level=info ts=2023-01-09T21:46:24.848572253Z caller=retention.go:47 msg="optional retention apply done"
level=info ts=2023-01-09T21:46:25.184809391Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=336.214489ms duration_ms=336 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:46:25.184934782Z caller=clean.go:34 msg="started cleaning of aborted partial uploads"
level=info ts=2023-01-09T21:46:25.184957378Z caller=clean.go:61 msg="cleaning of aborted partial uploads done"
level=info ts=2023-01-09T21:46:25.18497269Z caller=blocks_cleaner.go:44 msg="started cleaning of blocks marked for deletion"
level=info ts=2023-01-09T21:46:25.185006189Z caller=blocks_cleaner.go:58 msg="cleaning of blocks marked for deletion done"
level=info ts=2023-01-09T21:46:54.216566449Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=272.255358ms duration_ms=272 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:47:54.189112261Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=244.772637ms duration_ms=244 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:48:54.200035653Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=255.722853ms duration_ms=255 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:49:54.19546239Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=251.121839ms duration_ms=251 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:50:54.192010211Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=248.048608ms duration_ms=248 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:51:22.629114693Z caller=compact.go:1291 msg="start sync of metas"
level=info ts=2023-01-09T21:51:23.124510165Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=495.456617ms duration_ms=495 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:51:23.489003585Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=364.283994ms duration_ms=364 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:51:23.489160481Z caller=clean.go:34 msg="started cleaning of aborted partial uploads"
level=info ts=2023-01-09T21:51:23.489184234Z caller=clean.go:61 msg="cleaning of aborted partial uploads done"
level=info ts=2023-01-09T21:51:23.489199516Z caller=blocks_cleaner.go:44 msg="started cleaning of blocks marked for deletion"
level=info ts=2023-01-09T21:51:23.489234422Z caller=blocks_cleaner.go:58 msg="cleaning of blocks marked for deletion done"
level=info ts=2023-01-09T21:51:23.830168797Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=340.92912ms duration_ms=340 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:51:23.830249029Z caller=compact.go:1296 msg="start of GC"
level=info ts=2023-01-09T21:51:23.830808102Z caller=compact.go:1319 msg="start of compactions"
level=info ts=2023-01-09T21:51:23.831377604Z caller=compact.go:1355 msg="compaction iterations done"
level=info ts=2023-01-09T21:51:23.831422263Z caller=compact.go:430 msg="start first pass of downsampling"
level=info ts=2023-01-09T21:51:24.190375653Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=358.933946ms duration_ms=358 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:51:24.205484298Z caller=compact.go:444 msg="start second pass of downsampling"
level=info ts=2023-01-09T21:51:24.568955376Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=363.413786ms duration_ms=363 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:51:24.576929504Z caller=compact.go:451 msg="downsampling iterations done"
level=info ts=2023-01-09T21:51:24.965836982Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=388.865931ms duration_ms=388 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:51:24.965957173Z caller=retention.go:32 msg="start optional retention"
level=info ts=2023-01-09T21:51:24.966007935Z caller=retention.go:47 msg="optional retention apply done"
level=info ts=2023-01-09T21:51:25.316321705Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=350.29144ms duration_ms=350 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:51:25.316433468Z caller=clean.go:34 msg="started cleaning of aborted partial uploads"
level=info ts=2023-01-09T21:51:25.316456228Z caller=clean.go:61 msg="cleaning of aborted partial uploads done"
level=info ts=2023-01-09T21:51:25.31647158Z caller=blocks_cleaner.go:44 msg="started cleaning of blocks marked for deletion"
level=info ts=2023-01-09T21:51:25.316507383Z caller=blocks_cleaner.go:58 msg="cleaning of blocks marked for deletion done"
level=info ts=2023-01-09T21:51:54.204371029Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=260.059383ms duration_ms=260 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:52:54.227879132Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=283.528496ms duration_ms=283 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:53:54.190343766Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=246.472214ms duration_ms=246 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:54:54.211803818Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=267.022949ms duration_ms=267 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:55:54.221926716Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=277.071068ms duration_ms=277 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:56:22.62891162Z caller=compact.go:1291 msg="start sync of metas"
level=info ts=2023-01-09T21:56:23.114635293Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=485.717783ms duration_ms=485 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:56:23.465070627Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=350.239113ms duration_ms=350 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:56:23.465203722Z caller=clean.go:34 msg="started cleaning of aborted partial uploads"
level=info ts=2023-01-09T21:56:23.465227102Z caller=clean.go:61 msg="cleaning of aborted partial uploads done"
level=info ts=2023-01-09T21:56:23.465242744Z caller=blocks_cleaner.go:44 msg="started cleaning of blocks marked for deletion"
level=info ts=2023-01-09T21:56:23.465275876Z caller=blocks_cleaner.go:58 msg="cleaning of blocks marked for deletion done"
level=info ts=2023-01-09T21:56:23.819388416Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=354.104115ms duration_ms=354 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:56:23.81950765Z caller=compact.go:1296 msg="start of GC"
level=info ts=2023-01-09T21:56:23.820463874Z caller=compact.go:1319 msg="start of compactions"
level=info ts=2023-01-09T21:56:23.821316336Z caller=compact.go:1355 msg="compaction iterations done"
level=info ts=2023-01-09T21:56:23.82138269Z caller=compact.go:430 msg="start first pass of downsampling"
level=info ts=2023-01-09T21:56:24.168320702Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=346.903369ms duration_ms=346 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:56:24.177215784Z caller=compact.go:444 msg="start second pass of downsampling"
level=info ts=2023-01-09T21:56:24.549377418Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=372.125608ms duration_ms=372 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:56:24.559894033Z caller=compact.go:451 msg="downsampling iterations done"
level=info ts=2023-01-09T21:56:24.915826322Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=355.873806ms duration_ms=355 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:56:24.91594034Z caller=retention.go:32 msg="start optional retention"
level=info ts=2023-01-09T21:56:24.915990429Z caller=retention.go:47 msg="optional retention apply done"
level=info ts=2023-01-09T21:56:25.289759226Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=373.744093ms duration_ms=373 cached=170 returned=139 partial=0
level=info ts=2023-01-09T21:56:25.289893105Z caller=clean.go:34 msg="started cleaning of aborted partial uploads"
level=info ts=2023-01-09T21:56:25.289916592Z caller=clean.go:61 msg="cleaning of aborted partial uploads done"
level=info ts=2023-01-09T21:56:25.289932371Z caller=blocks_cleaner.go:44 msg="started cleaning of blocks marked for deletion"
level=info ts=2023-01-09T21:56:25.289969055Z caller=blocks_cleaner.go:58 msg="cleaning of blocks marked for deletion done"
level=info ts=2023-01-09T21:56:54.209820536Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=265.157385ms duration_ms=265 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:57:54.208502617Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=264.188245ms duration_ms=264 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:58:54.217343595Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=273.297876ms duration_ms=273 cached=170 returned=170 partial=0
level=info ts=2023-01-09T21:59:54.234353165Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=290.01077ms duration_ms=290 cached=170 returned=170 partial=0
level=info ts=2023-01-09T22:00:54.18465307Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=240.339876ms duration_ms=240 cached=170 returned=170 partial=0
```

</p>
</details>

**Anything else we need to know**:

It's possible a duplicate of https://github.com/thanos-io/thanos/issues/3394. It's similar but it seems the consensus is that it's due to an issue with alternative S3 implementations. In my case the S3 requests has not increased either, it's been stable for about 1 month.

**Comment by zekena2** [comment]
Same for me. I'm on version 0.29.0. we have a lot of 4xx errors (half of the requests to the s3 bucket are 4xx errors)

**Comment by albgus** [comment]
For what it's worth I seem to have been able to cut the amount of requests roughly in half by setting `--wait-interval=24h` and `--compact.cleanup-interval=0s`, drastically increasing the interval compactions are run. Specifically it seems the checks for no-compact.mark.json and deletion-mask.json that are much less frequent.

The `HEAD .../meta.json` requests are still very frequent though. It seems to kind of match up with this log message in the compactor logs:

```
level=info ts=2023-01-16T19:25:09.193425416Z caller=fetcher.go:478 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=292.429792ms duration_ms=292 cached=184 returned=184 partial=0
```

The message suggests that there is some form of cache, but perhaps this implementation is broken?

**Comment by yeya24** [comment]
The cache caches the metadata file itself, but Thanos still needs to call `HEAD .../meta.json` to check whether the metadata file still exists or not.

Tuning the interval could definititely help. Several flags you can tune:

```
--wait-interval=5m
--compact.cleanup-interval=5m
--compact.progress-interval=5m
--block-viewer.global.sync-block-interval=1m
```

If you don't need the compact progress metrics, you can simply set `compact.progress-interval` to 0.
If you don't need the compactor UI, you can simply set `web.disable` to false or increase value of `block-viewer.global.sync-block-interval`.
`--compact.cleanup-interval` is another value you could tune but I only recommend increasing the value but not set it to 0.
Increasing `--wait-interval` is also okay but usually I don't recommend changing it since it controls how often you compact.

**Comment by pjastrzabek-roche** [comment]
```
--compact.cleanup-interval is another value you could tune but I only recommend increasing the value but not set it to 0.
```
@yeya24 why is that?
when set to zero it still should be executed at the end of compaction loop. My reasoning was that I actually do not need all these things to be executed as soon as possible and when they run in parallel compactor component can have drastically different usage of cpu / memory so I prefer to have everything run in sequence.

**Comment by yeya24** [comment]
@pjastrzabek-roche That makes sense. leaning blocks at the end is okay, but that depends on your usecase.

**Comment by hicwic** [comment]
hi, 
facing this issue too.
we have a lot of 404 error retrieving meta datas from buckets.

Some additionnals infos that can help, maybe find a bug ?

Storegateway seems to be ok, fetching metas is only done one time every 3m (value of sync-block-duration)
Compactor have a weird number of fetching try, like 5 times the same meta for the same chunk every 5m

Example from my test env :
```
2023-08-29T11:31:45.408 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZTKCCY8PP5KH6KHT7TV43D/deletion-mark.json 10.42.5.5         706µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.411 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZWK3V1VGZMZPNR6FYCN2WC/deletion-mark.json 10.42.5.5         599µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.427 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZKQN520TJHSG0Y74VPNXW1/deletion-mark.json 10.42.5.5         750µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.440 [404 Not Found] s3.GetObject minio:9000/thanos/01H908ATX68JT4VPCRDDPWRX9P/deletion-mark.json 10.42.5.5         731µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.442 [404 Not Found] s3.GetObject minio:9000/thanos/01H8TQJ1WV80GJX1E08G6X75DG/deletion-mark.json 10.42.5.5         544µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.477 [404 Not Found] s3.GetObject minio:9000/thanos/01H901F3MTR28V0GTKET6PC0CH/deletion-mark.json 10.42.5.5         517µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.485 [404 Not Found] s3.GetObject minio:9000/thanos/01H908ATX68JT4VPCRDDPWRX9P/no-compact-mark.json 10.42.5.5         578µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.486 [404 Not Found] s3.GetObject minio:9000/thanos/01H8TQJ1WV80GJX1E08G6X75DG/no-compact-mark.json 10.42.5.5         429µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.488 [404 Not Found] s3.GetObject minio:9000/thanos/01H901F3MTR28V0GTKET6PC0CH/no-compact-mark.json 10.42.5.5         359µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.489 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZTKCCY8PP5KH6KHT7TV43D/no-compact-mark.json 10.42.5.5         351µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.490 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZWK3V1VGZMZPNR6FYCN2WC/no-compact-mark.json 10.42.5.5         404µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.491 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZKQN520TJHSG0Y74VPNXW1/no-compact-mark.json 10.42.5.5         590µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.508 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZTKCCY8PP5KH6KHT7TV43D/deletion-mark.json 10.42.5.5         651µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.510 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZWK3V1VGZMZPNR6FYCN2WC/deletion-mark.json 10.42.5.5         516µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.524 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZKQN520TJHSG0Y74VPNXW1/deletion-mark.json 10.42.5.5         650µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.531 [404 Not Found] s3.GetObject minio:9000/thanos/01H901F3MTR28V0GTKET6PC0CH/deletion-mark.json 10.42.5.5         640µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.557 [404 Not Found] s3.GetObject minio:9000/thanos/01H908ATX68JT4VPCRDDPWRX9P/deletion-mark.json 10.42.5.5         473µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.577 [404 Not Found] s3.GetObject minio:9000/thanos/01H8TQJ1WV80GJX1E08G6X75DG/deletion-mark.json 10.42.5.5         537µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.599 [404 Not Found] s3.GetObject minio:9000/thanos/01H901F3MTR28V0GTKET6PC0CH/no-compact-mark.json 10.42.5.5         728µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.602 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZTKCCY8PP5KH6KHT7TV43D/no-compact-mark.json 10.42.5.5         550µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.604 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZWK3V1VGZMZPNR6FYCN2WC/no-compact-mark.json 10.42.5.5         433µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.605 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZKQN520TJHSG0Y74VPNXW1/no-compact-mark.json 10.42.5.5         389µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.607 [404 Not Found] s3.GetObject minio:9000/thanos/01H908ATX68JT4VPCRDDPWRX9P/no-compact-mark.json 10.42.5.5         587µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.609 [404 Not Found] s3.GetObject minio:9000/thanos/01H8TQJ1WV80GJX1E08G6X75DG/no-compact-mark.json 10.42.5.5         412µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.664 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZKQN520TJHSG0Y74VPNXW1/deletion-mark.json 10.42.5.5         680µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.666 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZWK3V1VGZMZPNR6FYCN2WC/deletion-mark.json 10.42.5.5         486µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.700 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZTKCCY8PP5KH6KHT7TV43D/deletion-mark.json 10.42.5.5         615µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.709 [404 Not Found] s3.GetObject minio:9000/thanos/01H8TQJ1WV80GJX1E08G6X75DG/deletion-mark.json 10.42.5.5         400µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.714 [404 Not Found] s3.GetObject minio:9000/thanos/01H901F3MTR28V0GTKET6PC0CH/deletion-mark.json 10.42.5.5         618µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.716 [404 Not Found] s3.GetObject minio:9000/thanos/01H908ATX68JT4VPCRDDPWRX9P/deletion-mark.json 10.42.5.5         460µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.726 [404 Not Found] s3.GetObject minio:9000/thanos/01H908ATX68JT4VPCRDDPWRX9P/no-compact-mark.json 10.42.5.5         703µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.729 [404 Not Found] s3.GetObject minio:9000/thanos/01H901F3MTR28V0GTKET6PC0CH/no-compact-mark.json 10.42.5.5         596µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.730 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZWK3V1VGZMZPNR6FYCN2WC/no-compact-mark.json 10.42.5.5         467µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.732 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZKQN520TJHSG0Y74VPNXW1/no-compact-mark.json 10.42.5.5         500µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.734 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZTKCCY8PP5KH6KHT7TV43D/no-compact-mark.json 10.42.5.5         562µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.736 [404 Not Found] s3.GetObject minio:9000/thanos/01H8TQJ1WV80GJX1E08G6X75DG/no-compact-mark.json 10.42.5.5         478µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.761 [404 Not Found] s3.GetObject minio:9000/thanos/01H908ATX68JT4VPCRDDPWRX9P/deletion-mark.json 10.42.5.5         629µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.792 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZTKCCY8PP5KH6KHT7TV43D/deletion-mark.json 10.42.5.5         533µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.802 [404 Not Found] s3.GetObject minio:9000/thanos/01H8TQJ1WV80GJX1E08G6X75DG/deletion-mark.json 10.42.5.5         648µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.804 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZWK3V1VGZMZPNR6FYCN2WC/deletion-mark.json 10.42.5.5         1.291ms      ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.808 [404 Not Found] s3.GetObject minio:9000/thanos/01H901F3MTR28V0GTKET6PC0CH/deletion-mark.json 10.42.5.5         964µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.825 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZKQN520TJHSG0Y74VPNXW1/deletion-mark.json 10.42.5.5         544µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.833 [404 Not Found] s3.GetObject minio:9000/thanos/01H901F3MTR28V0GTKET6PC0CH/no-compact-mark.json 10.42.5.5         512µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.835 [404 Not Found] s3.GetObject minio:9000/thanos/01H8TQJ1WV80GJX1E08G6X75DG/no-compact-mark.json 10.42.5.5         518µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.842 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZWK3V1VGZMZPNR6FYCN2WC/no-compact-mark.json 10.42.5.5         584µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.844 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZKQN520TJHSG0Y74VPNXW1/no-compact-mark.json 10.42.5.5         508µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.846 [404 Not Found] s3.GetObject minio:9000/thanos/01H908ATX68JT4VPCRDDPWRX9P/no-compact-mark.json 10.42.5.5         867µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.848 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZTKCCY8PP5KH6KHT7TV43D/no-compact-mark.json 10.42.5.5         355µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.867 [404 Not Found] s3.GetObject minio:9000/thanos/01H8TQJ1WV80GJX1E08G6X75DG/deletion-mark.json 10.42.5.5         546µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.872 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZKQN520TJHSG0Y74VPNXW1/deletion-mark.json 10.42.5.5         573µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.874 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZWK3V1VGZMZPNR6FYCN2WC/deletion-mark.json 10.42.5.5         443µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.875 [404 Not Found] s3.GetObject minio:9000/thanos/01H908ATX68JT4VPCRDDPWRX9P/deletion-mark.json 10.42.5.5         448µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.893 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZTKCCY8PP5KH6KHT7TV43D/deletion-mark.json 10.42.5.5         577µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.897 [404 Not Found] s3.GetObject minio:9000/thanos/01H901F3MTR28V0GTKET6PC0CH/deletion-mark.json 10.42.5.5         717µs       ↑ 93 B ↓ 704 B
2023-08-29T11:31:45.936 [404 Not Found] s3.GetObject minio:9000/thanos/01H8TQJ1WV80GJX1E08G6X75DG/no-compact-mark.json 10.42.5.5         651µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.938 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZKQN520TJHSG0Y74VPNXW1/no-compact-mark.json 10.42.5.5         489µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.945 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZWK3V1VGZMZPNR6FYCN2WC/no-compact-mark.json 10.42.5.5         710µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.947 [404 Not Found] s3.GetObject minio:9000/thanos/01H908ATX68JT4VPCRDDPWRX9P/no-compact-mark.json 10.42.5.5         710µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.949 [404 Not Found] s3.GetObject minio:9000/thanos/01H8ZTKCCY8PP5KH6KHT7TV43D/no-compact-mark.json 10.42.5.5         522µs       ↑ 93 B ↓ 708 B
2023-08-29T11:31:45.957 [404 Not Found] s3.GetObject minio:9000/thanos/01H901F3MTR28V0GTKET6PC0CH/no-compact-mark.json 10.42.5.5         620µs       ↑ 93 B ↓ 708 B
```

Debug log of my compactor (we see 5 times "fetching metadata")
```
ts=2023-08-29T09:31:45.289526676Z caller=compact.go:1414 level=info msg="start sync of metas"
ts=2023-08-29T09:31:45.289622645Z caller=fetcher.go:317 level=debug component=block.BaseFetcher msg="fetching meta data" concurrency=1
ts=2023-08-29T09:31:45.493687261Z caller=fetcher.go:487 level=info component=block.BaseFetcher msg="successfully synchronized block metadata" duration=204.113823ms duration_ms=204 cached=20 returned=6 partial=0
ts=2023-08-29T09:31:45.493784593Z caller=compact.go:1419 level=info msg="start of GC"
ts=2023-08-29T09:31:45.493991947Z caller=compact.go:1442 level=info msg="start of compactions"
ts=2023-08-29T09:31:45.494351035Z caller=fetcher.go:317 level=debug component=block.BaseFetcher msg="fetching meta data" concurrency=1
ts=2023-08-29T09:31:45.494846981Z caller=compact.go:1478 level=info msg="compaction iterations done"
ts=2023-08-29T09:31:45.494909438Z caller=compact.go:457 level=info msg="downsampling was explicitly disabled"
ts=2023-08-29T09:31:45.610485595Z caller=fetcher.go:487 level=info component=block.BaseFetcher msg="successfully synchronized block metadata" duration=116.27933ms duration_ms=116 cached=20 returned=6 partial=0
ts=2023-08-29T09:31:45.610579658Z caller=clean.go:34 level=info msg="started cleaning of aborted partial uploads"
ts=2023-08-29T09:31:45.610593516Z caller=clean.go:61 level=info msg="cleaning of aborted partial uploads done"
ts=2023-08-29T09:31:45.610600641Z caller=blocks_cleaner.go:44 level=info msg="started cleaning of blocks marked for deletion"
ts=2023-08-29T09:31:45.610617154Z caller=blocks_cleaner.go:58 level=info msg="cleaning of blocks marked for deletion done"
ts=2023-08-29T09:31:45.61072699Z caller=fetcher.go:317 level=debug component=block.BaseFetcher msg="fetching meta data" concurrency=1
ts=2023-08-29T09:31:45.737532847Z caller=fetcher.go:487 level=info component=block.BaseFetcher msg="successfully synchronized block metadata" duration=126.899903ms duration_ms=126 cached=20 returned=6 partial=0
ts=2023-08-29T09:31:45.737742783Z caller=fetcher.go:317 level=debug component=block.BaseFetcher msg="fetching meta data" concurrency=1
ts=2023-08-29T09:31:45.849343472Z caller=fetcher.go:487 level=info component=block.BaseFetcher msg="successfully synchronized block metadata" duration=111.628955ms duration_ms=111 cached=20 returned=6 partial=0
ts=2023-08-29T09:31:45.849412322Z caller=retention.go:32 level=info msg="start optional retention"
ts=2023-08-29T09:31:45.849424915Z caller=retention.go:47 level=info msg="optional retention apply done"
ts=2023-08-29T09:31:45.849479106Z caller=fetcher.go:317 level=debug component=block.BaseFetcher msg="fetching meta data" concurrency=1
ts=2023-08-29T09:31:45.959667663Z caller=fetcher.go:487 level=info component=block.BaseFetcher msg="successfully synchronized block metadata" duration=110.229402ms duration_ms=110 cached=20 returned=6 partial=0
ts=2023-08-29T09:31:45.959738949Z caller=clean.go:34 level=info msg="started cleaning of aborted partial uploads"
ts=2023-08-29T09:31:45.959764104Z caller=clean.go:61 level=info msg="cleaning of aborted partial uploads done"
ts=2023-08-29T09:31:45.959773644Z caller=blocks_cleaner.go:44 level=info msg="started cleaning of blocks marked for deletion"
ts=2023-08-29T09:31:45.959790795Z caller=blocks_cleaner.go:58 level=info msg="cleaning of blocks marked for deletion done"
```

Digging the code since yesterday, and cannot find why this call is duplicated many times

have you any idea about this issue ?

**Comment by thorker** [comment]
I think it's a combination of different parts of a single compaction run(compact, downsample, delete) and the additional phases(progress metrics, cleanup & block sync for the viewer). In each of this runs or phases, the component block.Basefetcher checks for the marker files above (no-compact-mark.json,deletion-mark.json,no-downsample-mark.json) and that results in the high api calls to the storage.

**Comment by julienlau** [comment]
same issue here on a bare metal env.
Thanos is rate limited by the S3 store.

**Comment by manavadariakevin** [comment]
Facing similar issue in thanos compator and i am getting below URIs very frequently. 
deletion-mark.json
no-compact-mark.json
no-downsample-mark.json

Even after applying above changes , it is not helping 
here is my conf and also showing the logs 

    extraFlags:
      - "--delete-delay=1h"
      - "--compact.blocks-fetch-concurrency=10"
      - "--compact.concurrency=10"
      - "--compact.enable-vertical-compaction"
      - "--block-files-concurrency=20"
      # - "--compact.skip-block-with-out-of-order-chunks"
      - "--block-meta-fetch-concurrency=32"
      - "--downsample.concurrency=50"
      - "--block-discovery-strategy=recursive"
      - "--deduplication.replica-label=replica"
      - "--deduplication.func=penalty"
      - "--wait-interval=24h"
      - "--compact.cleanup-interval=0s"

![image](https://github.com/user-attachments/assets/93c61124-cc4e-462c-9ac9-dd0bad541160)

**Comment by aaron-burt-murine** [comment]
I found that adding `--block-discovery-strategy=recursive` reduced S3 API calls to reasonable levels.

**Comment by julienlau** [comment]
> Facing similar issue in thanos compator and i am getting below URIs very frequently. deletion-mark.json no-compact-mark.json no-downsample-mark.json
> 
> Even after applying above changes , it is not helping here is my conf and also showing the logs
> 
> ```
> extraFlags:
>   - "--delete-delay=1h"
>   - "--compact.blocks-fetch-concurrency=10"
>   - "--compact.concurrency=10"
>   - "--compact.enable-vertical-compaction"
>   - "--block-files-concurrency=20"
>   # - "--compact.skip-block-with-out-of-order-chunks"
>   - "--block-meta-fetch-concurrency=32"
>   - "--downsample.concurrency=50"
>   - "--block-discovery-strategy=recursive"
>   - "--deduplication.replica-label=replica"
>   - "--deduplication.func=penalty"
>   - "--wait-interval=24h"
>   - "--compact.cleanup-interval=0s"
> ```
> 
> ![image](https://github.com/user-attachments/assets/93c61124-cc4e-462c-9ac9-dd0bad541160)

the option --wait-interval is ignored if the option --wait is not set. Please check the doc

**Comment by bumarcell** [comment]
For those still hitting this: most of the compactor's S3 calls in our setup were the marker probes. Every metadata sync GETs `deletion-mark.json`, `no-compact-mark.json` and `no-downsample-mark.json` for every block, and since markers are rare nearly all of those come back 404. On a bucket with ~4.5k live blocks that was ~13k GETs per sync and 45-230 req/s depending on how often the compaction, downsampling, retention and progress loops synced.

`--block-discovery-strategy=recursive` already lists every object name (and removes the per-block `HEAD meta.json`), but the marker filters ignored that and kept probing. #9025 makes the recursive lister record the marker files it sees and lets the three filters skip blocks whose marker was not listed, so the marker 404s disappear entirely with the recursive strategy. Nothing changes for the default `concurrent` strategy. Feedback from anyone who can test it against their store is welcome.
